docs/coding-guidelines.md

```diff
@@ -0,0 +1,403 @@
+
+### Argument null checking
+
+Null checking is required for parameters that cannot be null (big surprise!). All of our null checks are within the method body.
+In constructors where a field or property is initialized coalescing checks are allowed. Otherwise use the regular bracing approach.
+
+```cs
+
+    Path = path ?? throw new ArgumentNullException(nameof(path));
+    ....
+    if (item == null)
+    {
+        throw new ArgumentNullException(nameof(item));
+    }
+
+```
+
+### Optional parameters
+
+Optional parameters are allowed in certain situations, but discouraged.
+Optional parameters are at the root for many versioning issues.
+The optional parameter default values are a compile time decision, that's why adding a new optional parameter to an API is a breaking change!
+Furthermore, if you have a method with optional parameters, adding an overload with additional optional parameters might cause a compile-time breaking change.
+Optional parameters in internal/private methods are acceptable.
+
+### Async method patterns
+
+By default all async methods must have the `Async` suffix. There are some exceptional circumstances where a method name from a previous framework will be grandfathered in.
+
+Passing cancellation tokens can be passed an optional parameter with a value of `default(CancellationToken)`, which is equivalent to `CancellationToken.None`.
+Prefer not having optional parameters.
+
+Sample async method:
+
+```cs
+public Task GetDataAsync(
+    QueryParams query,
+    int maxData,
+    CancellationToken cancellationToken)
+{
+    ...
+}
+```
+
+### Extension method patterns
+
+The general rule is: if a regular static method would suffice, avoid extension methods.
+
+Extension methods are often useful to create chainable method calls, for example, when constructing complex objects, or creating queries.
+Internal extension methods are allowed, but bear in mind the previous guideline: ask yourself if an extension method is truly the most appropriate pattern.
+The namespace of the extension method class should generally be the namespace that represents the functionality of the extension method, as opposed to the namespace of the target type.
+The class name of an extension method container (also known as a "sponsor type") should generally follow the pattern of `<Feature>Extensions`, `<Target><Feature>Extensions`, or `<Feature><Target>Extensions`. For example:
+
+```cs
+namespace Food
+{
+    class Fruit { ... }
+}
+namespace Fruit.Eating
+{
+    class FruitExtensions { public static void Eat(this Fruit fruit); }
+  OR
+    class FruitEatingExtensions { public static void Eat(this Fruit fruit); }
+  OR
+    class EatingFruitExtensions { public static void Eat(this Fruit fruit); }
+}
+```
+
+When writing extension methods for an interface the sponsor type name must not start with an `I`.
+
+### Doc comments
```

Comment:
Does this refer to XMLDoc, or docs.microsoft.com?

Reply:
Fixed it to say Code comments.